packages/fmg-nitro-adjudicator/contracts/NitroAdjudicator.sol

```diff
@@ -0,0 +1,420 @@
+        uint[] amount;
+        uint256 finalizedAt;
+        State.StateStruct challengeState;
+    }
+    struct Guarantee {
+        address guarantor;
+        address target;
+        address[] priorities;
+    }
+    struct Signature {
+        uint8 v;
+        bytes32 r;
+        bytes32 s;
+    }
+    struct ConclusionProof {
+        State.StateStruct penultimateState;
+        Signature penultimateSignature;
+        State.StateStruct ultimateState;
+        Signature ultimateSignature;
+    }
+
+    mapping(address => uint) public allocations;
+    mapping(address => Outcome) public outcomes;
+
+    // TODO: Challenge duration should depend on the channel
+    uint constant CHALLENGE_DURATION = 5 minutes;
+
+    // **************
+    // Eth Management
+    // **************
+
+    function deposit(address destination) public payable {
+        allocations[destination] = allocations[destination] + msg.value;
+    }
+
+    function withdraw(address participant, address payable destination, uint amount, uint8 _v, bytes32 _r, bytes32 _s) public payable {
+        require(
+            allocations[participant] >= amount,
+            "Withdraw: overdrawn"
+        );
+        Authorization memory authorization = Authorization(
+            participant,
+            destination,
+            amount,
+            msg.sender
+        );
+
+        require(
+            recoverSigner(abi.encode(authorization), _v, _r, _s) == participant,
+            "Withdraw: not authorized by participant"
+        );
+
+        allocations[participant] = allocations[participant] - amount;
+        destination.transfer(amount);
+    }
+
+    function transfer(address channel, address destination, uint amount) public {
+        require(
+            outcomes[channel].finalizedAt < now,
+            "Transfer: outcome must be final"
+        );
+        require(
+            outcomes[channel].finalizedAt > 0,
+            "Transfer: outcome must be present"
+        );
+
+        uint owedToDestination = overlap(destination, outcomes[channel], amount);
+
+        require(
+            owedToDestination <= allocations[channel],
+            "Transfer: allocations[channel] must cover transfer"
+        );
+        require(
+            owedToDestination >= amount,
+            "Transfer: transfer too large"
+        );
+
+        allocations[destination] = allocations[destination] + owedToDestination;
+        allocations[channel] = allocations[channel] - owedToDestination;
+
+        outcomes[channel] = remove(outcomes[channel], destination, amount);
+    }
+
+    function claim(address recipient, Guarantee memory guarantee, uint amount) public {
+        require(
+            isChannelClosed(guarantee.target),
+            "Claim: channel must be closed"
+        );
+        require(
+            // the outcome is assumed to be valid, so this check is sufficient
+            guarantee.priorities.length == outcomes[guarantee.target].destination.length,
```

Comment:
The priorities list isn't intended to completely overlap the target outcomes. The rule for priorities is that if an outcome address appears in the priorities it is boosted to the front. It's possible to have entries in the priorities list that don't appear in the outcome and visa-versa.